From 378ebf254a81f3fd66ed1abd99da1168e18693c2 Mon Sep 17 00:00:00 2001
From: Xiaofeng Wei <xiaofeng.wei@nxp.com>
Date: Wed, 5 Mar 2025 16:04:57 +0800
Subject: [PATCH 48/51] arm64: dts: Add imx91-11x11-frdm-storm-mt9m114.dts

Add parallel interface camera MT9M114 support for FRDM-IMX91 Storm board.

Signed-off-by: Xiaofeng Wei <xiaofeng.wei@nxp.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |   3 +-
 .../imx91-11x11-frdm-storm-mt9m114.dts        | 173 ++++++++++++++++++
 2 files changed, 175 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm64/boot/dts/freescale/imx91-11x11-frdm-storm-mt9m114.dts

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 5d15a6ffebdf..d31f4a49175c 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -417,7 +417,8 @@ dtb-$(CONFIG_ARCH_MXC) += imx91-11x11-evk.dtb \
 			  imx91-11x11-frdm-storm-aud-hat.dtb \
 			  imx91-11x11-frdm-storm-8mic.dtb \
 			  imx91-11x11-frdm-storm-tianma-wvga-panel.dtb \
-			  imx91-11x11-frdm-storm-fec.dtb
+			  imx91-11x11-frdm-storm-fec.dtb \
+			  imx91-11x11-frdm-storm-mt9m114.dtb
 
 imx93-11x11-evk-pmic-pf0900-dtbs := imx93-11x11-evk.dtb imx93-11x11-evk-pmic-pf0900.dtbo
 imx93-11x11-evk-pmic-pf0900-root-dtbs := imx93-11x11-evk-root.dtb imx93-11x11-evk-pmic-pf0900.dtbo
diff --git a/arch/arm64/boot/dts/freescale/imx91-11x11-frdm-storm-mt9m114.dts b/arch/arm64/boot/dts/freescale/imx91-11x11-frdm-storm-mt9m114.dts
new file mode 100644
index 000000000000..ad9b78250cd2
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx91-11x11-frdm-storm-mt9m114.dts
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2025 NXP
+ */
+
+#include "imx91-11x11-frdm-storm.dts"
+
+/ {
+	reg_mt9m114_dovdd: regulator-mt9m114-dovdd {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-dovdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		startup-delay-us = <0>;
+		off-on-delay-us  = <1000>;
+		gpio = <&pca9538 0 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_exp_3v3>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_avdd: regulator-mt9m114-avdd {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-avdd";
+		regulator-min-microvolt = <2800000>;
+		regulator-max-microvolt = <2800000>;
+		startup-delay-us = <200>;
+		off-on-delay-us  = <500>;
+		gpio = <&pca9538 1 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_exp_3v3>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_dvdd: regulator-mt9m114-dvdd {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-dvdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		startup-delay-us = <500>;
+		off-on-delay-us  = <200>;
+		gpio = <&pca9538 2 GPIO_ACTIVE_HIGH>;
+		vin-supply = <&reg_exp_3v3>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_extclk: regulator-mt9m114-extclk {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-extclk";
+		startup-delay-us = <10000>;
+		off-on-delay-us  = <200>;
+		gpio = <&pca9538 3 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	reg_mt9m114_ctrl_9509: regulator-mt9m114-ctrl-9509 {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-ctrl-9509";
+		gpio = <&pca9538 4 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-always-on;
+	};
+
+	reg_mt9m114_ctrl_4t245: regulator-mt9m114-ctrl-4t245 {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-ctrl-4t245";
+		startup-delay-us = <10000>;
+		off-on-delay-us  = <200>;
+		gpio = <&pca9538 5 GPIO_ACTIVE_HIGH>;
+		enable-active-low;
+	};
+
+	reg_mt9m114_ctrl_16t245: regulator-mt9m114-ctrl-16t245 {
+		compatible = "regulator-fixed";
+		regulator-name = "mt9m114-ctrl-16t245";
+		gpio = <&pca9538 6 GPIO_ACTIVE_HIGH>;
+		enable-active-low;
+		regulator-always-on;
+	};
+};
+
+&iomuxc {
+	pinctrl_parallel_csi: ci_pi {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO01__MEDIAMIX_CAM_DATA0		0xb9e
+			MX91_PAD_GPIO_IO07__MEDIAMIX_CAM_DATA1		0xb9e
+			MX91_PAD_GPIO_IO08__MEDIAMIX_CAM_DATA2		0xb9e
+			MX91_PAD_GPIO_IO09__MEDIAMIX_CAM_DATA3		0xb9e
+			MX91_PAD_GPIO_IO10__MEDIAMIX_CAM_DATA4		0xb9e
+			MX91_PAD_GPIO_IO11__MEDIAMIX_CAM_DATA5		0xb9e
+			MX91_PAD_GPIO_IO14__MEDIAMIX_CAM_DATA6		0xb9e
+			MX91_PAD_GPIO_IO15__MEDIAMIX_CAM_DATA7		0xb9e
+			MX91_PAD_GPIO_IO17__MEDIAMIX_CAM_DATA8		0xb9e
+			MX91_PAD_GPIO_IO18__MEDIAMIX_CAM_DATA9		0xb9e
+
+			MX91_PAD_GPIO_IO00__MEDIAMIX_CAM_CLK		0xb9e
+			MX91_PAD_GPIO_IO02__MEDIAMIX_CAM_VSYNC		0xb9e
+			MX91_PAD_GPIO_IO03__MEDIAMIX_CAM_HSYNC		0xb9e
+			MX91_PAD_GPIO_IO05__GPIO2_IO5			0x31e
+			MX91_PAD_GPIO_IO04__GPIO2_IO4			0x31e
+		>;
+	};
+
+	pinctrl_lpi2c8: lpi2c8grp {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO12__LPI2C8_SDA		0x40000b9e
+			MX91_PAD_GPIO_IO13__LPI2C8_SCL		0x40000b9e
+		>;
+	};
+};
+
+&lpi2c1 {
+	/delete-node/camera@3c;
+};
+
+&lpuart3 {
+	status = "disabled";
+};
+
+&parallel_csi {
+	status = "okay";
+
+	/delete-node/port;
+	port {
+		parallel_csi_ep: endpoint {
+			remote-endpoint = <&mt9m114_ep>;
+		};
+	};
+};
+
+&lpi2c8 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	clock-frequency = <400000>;
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&pinctrl_lpi2c8>;
+	pinctrl-1 = <&pinctrl_lpi2c8>;
+	status = "okay";
+
+	pca9538: gpio@70 {
+		compatible = "nxp,pca9538";
+		reg = <0x70>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		vcc-supply = <&reg_exp_3v3>;
+		status = "okay";
+	};
+
+	mt9m114: mt9m114@48 {
+		compatible = "on,mt9m114";
+		reg = <0x48>;
+		pinctrl-names    = "default";
+		pinctrl-0        = <&pinctrl_parallel_csi>;
+		reset-gpios      = <&gpio2 4 GPIO_ACTIVE_HIGH>;
+		powerdown-gpios  = <&gpio2 5 GPIO_ACTIVE_HIGH>;
+		DOVDD-supply     = <&reg_mt9m114_dovdd>;
+		AVDD-supply      = <&reg_mt9m114_avdd>;
+		DVDD-supply      = <&reg_mt9m114_dvdd>;
+		EXTCLK-supply    = <&reg_mt9m114_extclk>;
+		CTRL_4T245-supply    = <&reg_mt9m114_ctrl_4t245>;
+		mclk             = <27000000>;
+		status = "okay";
+
+		port {
+			mt9m114_ep: endpoint {
+				remote-endpoint = <&parallel_csi_ep>;
+				bus-type = <5>; /* V4L2_FWNODE_BUS_TYPE_PARALLEL */
+				bus-width = <8>;
+				vsync-active = <1>;
+				hsync-active = <1>;
+				pclk-sample = <1>;
+			};
+		};
+	};
+};
-- 
2.34.1

